feat: find closest pairs using tree-sitter

woojiq opened the pull request:
Closes #8325 
closes  #3615
closes #3432
closes #4475
closes #7012

Finding the closest surround pairs (`mim`, `mam`) was done without being aware of the syntax. This PR adds a tree-sitter implementation that may be faster (untested), but most important it's more correct.
If the syntax distinct strings than `mim` and `mam` will work inside strings (didn't work with plaintext impl), matching should also work inside injections.

<details>
<summary>OLD</summary>
`find_nth_closest_pairs_pos` produces unsorted tuple. The tuple must be sorted before passing it to the change Transaction. Otherwise, this statement will fail:
https://github.com/helix-editor/helix/blob/13d4463e4146473391e37b7f75e99b731aa55878/helix-core/src/transaction.rs#L582-L585

MRE: 
1. `hx`
2. repeat exact keys (two spaces after "("): `(  <esc>hvhmdm`. Make sure `auto-pairs` is turn on.
</details>

woojiq commented:
I've also added a commit that fixes another bug that's not really related to this one. But I found it while debugging crash. So I thought it would be good to fix both 😃.

<hr/>

Message from second commit: [fix: preserve selection direction when searching for char pair](https://github.com/helix-editor/helix/pull/8294/commits/2674cafc3922c13afa624b3a85e6cd538103c230) 

`find_nth_closests_pairs_pos` checks the direction backward or forward. This helps preserve the selection direction when pressing `mim` or `mam`. This commit also adds this functionality to `find_nth_pairs_pos`, so `mi(` or `ma}` also preserves direction now.

<hr/>

I know it's bad practice to include different types of commits in one PR, but they are really small (commits are 5 lines + 20 line tests). I've written a nice commit description for each, so you don't need to squash.

the-mikedavis commented:
I think it makes sense historically: `surround_*` functions were all introduced together before textobjects: #320 vs #728, but I think it makes more sense to use tree-sitter if possible (ideally the same code path as `mm`?)

woojiq commented:
> mim/mam (which uses TS).

Under the hood, both commands use https://github.com/helix-editor/helix/blob/2284bce970e9175fd81851e1616218a6119354fc/helix-term/src/commands.rs#L5163 which uses function that doesn't use TS https://github.com/helix-editor/helix/blob/2284bce970e9175fd81851e1616218a6119354fc/helix-core/src/textobject.rs#L230-L230

Though, as @<!-- -->the-mikedavis mentioned, `mm` uses TS to find the matching brackets https://github.com/helix-editor/helix/blob/2284bce970e9175fd81851e1616218a6119354fc/helix-core/src/match_brackets.rs#L54 I think we need to rewrite `find_nth_closest_pairs_pos` to use/reuse something like `find_pair` (but n times) when TS is present and fall back to the current implementation otherwise. Is it a good one?

pascalkuthe reviewed (commented):
I still need to do some testing but from reading trough the changes this looks great.

pascalkuthe commented on helix-core/src/surround.rs:
```diff
@@ -52,7 +57,58 @@ pub fn get_pair(ch: char) -> (char, char) {
+    range: Range,
+    skip: usize,
+) -> Result<(usize, usize)> {
+    let is_close_pair = |ch| PAIRS.iter().any(|(_, close)| *close == ch);
```
I think autopairs could stay separate for now.

For example.ple we don't want to add `'` there since that I'd also often used as standalone (like rust lifetimes).

Although I guess that could also be handled by language co fig (autpais are co figural iirc). Thoughts @the-mikedavis

the-mikedavis replied:
Yeah let's keep auto-pairs separate for now, those need some special logic for things like `'` as Pascal says

woojiq replied:
Done in 938222478301233fd7328ec2aa59a33381e35060
No new functionality added, only moved PAIRS-related stuff to `match_bracking.rs`. I think it's a better place than `surround.rs` or `helix_core::lib.rs`.

pascalkuthe reviewed (dismissed):
Changes look go to me. I will do some testing in my daily driver branch (I use mim/mam quite a bit)

woojiq commented:
I found one discrepancy. Since I combined the pairs into one list, quotes are now considered as pairs too. But if you use a file without the tree-sitter syntax, it will still skip `mim|mam` for quotes:
Try  `mim` on this text `{ " cursor here " }` because current impl of nearest brackets first checks if the char is open and continues to look for a closed char if it is false.
https://github.com/helix-editor/helix/blob/9f6e9a151228c3c95cddb9032849625883d2ac8a/helix-core/src/surround.rs#L70
The code used to work correctly because we had a different set of pairs (without the quotes). But now that the pairs are merged into one, I believe it should also accept quotes as pairs (they do with tree-sitter, this "bug" only applies to non-tree-sitter files). The solution is simple - give the check of the closing element a higher priority. Do I need to do this?

pascalkuthe commented:
Hmm good qustion we might wnat to just not support quotes for plaintext files since its not clear whether we are currently "inside" the pair or not. We do the same for match_bracket.

woojiq commented:
>  we might wnat to just not support quotes for plaintext

If so, the code now works properly, but for clarity I might add something like `if is_quote(ch)` to the plain implementation to make it more obvious to readers what's going on.

pascalkuthe commented:
> > we might wnat to just not support quotes for plaintext
> 
> If so then it works as expected, but for clarity I might add something like `if is_quote(ch)` to the plain implementation to make it more obvious to readers what's going on.

I think a comment would be good :+1:

woojiq commented:
So I took a closer look at the behavior of the code with gdb and found that this is not true:
> We do the same for match_bracket.

On the master, it's still going through all the text trying to find closing quote. But it just never finds it, because there is no opposite state for quotes. I don't want to do it in this PR, but I'll probably make another one when I have some free time to clean up match_brackets.rs and surround.rs a bit.
A few things need to be done:
1. Name functions and variables accordingly. Some have "bracket" in the name but actually check all pairs, etc.
2. Distinguish between quotes and brackets so there is no hidden flows when matching on plaintext.

pascalkuthe commented:
hmm I think you are right reading the code again. I think that is an oversight tough the original PR stated it worked that way.

I think I assumed from the naming of `is_valid_bracket` that it would exclude quotes (and therefor bail here: https://github.com/helix-editor/helix/blob/9f6e9a151228c3c95cddb9032849625883d2ac8a/helix-core/src/match_brackets.rs#L145) during review so it would be good to clean that up.

It doesn't actually work tough if you test it (pressing mm on quotes in plaintext files does nothing).

pascalkuthe commented:
would really love to see this land and think its almost there. Are you interested in finishing this @woojiq?

woojiq commented:
> would really love to see this land and think its almost there. Are you interested in finishing this @woojiq?

Sure, afair this pr is almost ready, I'm going to look again and rebase in 1-2 days.

woojiq commented:
I've rebased and changed this PR a bit. To me it looks much better than before. I wrote few tests for `mam` (logic that changes selection based on the mod `Around | Inside` wasn't changed.

pascalkuthe commented on helix-core/src/match_brackets.rs:
```diff
@@ -9,16 +9,32 @@ use crate::Syntax;
+// The difference between BRACKETS and PAIRS is that we can find matching
+// BRACKETS in a plain text file, but we can't do the same for PAIRs.
+// PAIRS also contains all BRACKETS.
+pub const PAIRS: [(char, char); BRACKETS.len() + 3] = {
+    let mut pairs = [(' ', ' '); BRACKETS.len() + 3];
+    let mut idx = 0;
+    while idx < BRACKETS.len() {
+        pairs[idx] = BRACKETS[idx];
+        idx += 1;
+    }
+    pairs[idx] = ('"', '"');
+    pairs[idx + 1] = ('\'', '\'');
+    pairs[idx + 2] = ('`', '`');
+    pairs
+};
```
hopefully more std functions become const so we can use copy_from_slice here for exampple

pascalkuthe commented on helix-term/tests/test/commands.rs:
```diff
@@ -640,3 +640,63 @@ async fn test_join_selections_space() -> anyhow::Result<()> {
     Ok(())
 }
+
+#[tokio::test(flavor = "multi_thread")]
```
thanks for writing tests!

pascalkuthe reviewed (approved):
thanks for finishing that this will be great to have!

the-mikedavis reviewed (dismissed):
I agree, looks great! Could you rebase / merge master to resolve the conflicts?

thomasaarholt commented:
@woojiq, I found that this doesn't work for quotation symbols (") in Python, and only on the _opening_ quotation mark in Rust (for `let foo = "hello world"`, `mim` only works with the cursor positioned on the first `"`, not the last one).
[See my comment here for more details](https://github.com/helix-editor/helix/issues/4475#issuecomment-2079350150).

pascalkuthe commented:
https://github.com/helix-editor/helix/issues/4475#issuecomment-2079372893

Only the rudt case is abug either this PR, the pythpm case is an existing bug with `mm` that I wanted to fix for a while and you don't need to look into.
